### PR TITLE
Introduce ZiplineServiceType

### LINKED ISSUE #556: ZiplineClass
I’m not sure of the name. We should consider adding a type to reflectively describe a `ZiplineService`.

```
interface ZiplineClass {
  val functions: List<ZiplineFunction>
}
```

Then we can have this:
```
class Zipline {
  val services: Map<String, ZiplineClass>
  val clients: Map<String, ZiplineClass>
}
```

One challenge: would we serialize the `ZiplineFunction` instances so they always described the inbound side? This is interesting if inbound and outbound have different functions for the same service.

### PR DESCRIPTION
**Pull request by swankjesse**
This is a new reflection-like API to introspect what functions are available on the other side of the bridge.

Working towards this: https://github.com/cashapp/zipline/issues/645

Closes: https://github.com/cashapp/zipline/issues/556

### CODE REVIEW

File: zipline/src/hostTest/kotlin/app/cash/zipline/ZiplineServiceTypeTest.kt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    val functions = type.functions.sortedBy { it.name }
+    assertEquals(2, functions.size)
+
+    assertEquals("fun close(): kotlin.Unit", functions[0].name)
+    assertFalse(functions[0].isSuspending)
+    assertTrue(functions[0].isClose)
+
+    assertEquals(
+      "fun echo(app.cash.zipline.testing.EchoRequest): app.cash.zipline.testing.EchoResponse",
```
Comment (swankjesse) [inline zipline/src/hostTest/kotlin/app/cash/zipline/ZiplineServiceTypeTest.kt:62]: I haven’t yet taken the next step of making this accessible through the public API.

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/Endpoint.kt [hunk 2]
```diff
@@ -112,7 +111,7 @@ class Endpoint internal constructor(
     }
   }
 
-  private val functionsCache = mutableMapOf<String, List<ZiplineFunction<*>>>()
+  private val serviceTypeCache = mutableMapOf<String, RealZiplineServiceType<*>>()
```
Comment (swankjesse) [inline zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/Endpoint.kt:114]: I’m changing the cache from `List<ZiplineFunction>` to `ZiplineServiceTypes`. Big benefit there is we don’t need to build the HashMap organized by name every single time.

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/Endpoint.kt [hunk 3]
```diff
@@ -153,8 +146,8 @@ class Endpoint internal constructor(
     // Detect leaked old services when creating new services.
     detectLeaks()
 
-    val functions = adapter.ziplineFunctions(json.serializersModule)
-    val callHandler = OutboundCallHandler(name, this, adapter, scope, functions)
+    val type = serviceType(adapter)
```
Comment (swankjesse) [inline zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/Endpoint.kt:149]: Now we’re caching on both sides of the endpoint.

**Review by JakeWharton — APPROVED** [review]
Good red bits

**Pull request merged**